src/lightarti/tor-dirmgr/src/state.rs

```diff
@@ -167,18 +195,45 @@ impl<DM: WriteNetDir> GetConsensusState<DM> {
         &mut self,
         from_cache: bool,
         text: &str,
+        churn: Vec<RsaIdentity>,
     ) -> Result<Option<&ConsensusMeta>> {
         // Try to parse it and get its metadata.
-        let (consensus_meta, unvalidated) = {
-            let (signedval, remainder, parsed) = MdConsensus::parse(text)?;
+        let (consensus_meta, mut unvalidated) = {
+            let (signed_part, remainder, parsed) = MdConsensus::parse(text)?;
             if let Ok(timely) = parsed.check_valid_now() {
-                let meta = ConsensusMeta::from_unvalidated(signedval, remainder, &timely);
+                let meta = ConsensusMeta::from_unvalidated(signed_part, remainder, &timely);
                 (meta, timely)
             } else {
                 return Ok(None);
             }
         };
 
+        // Filter out the churn.
+        if ! churn.is_empty() {
+            let mut churn_set: HashSet<&RsaIdentity> = HashSet::from_iter(&churn);
+
+            // If the churn is above a threshold, we only consider a random subset
+            // of the churned routers.
+            let churn_threshold = unvalidated.consensus.routers.len() / 6;
+            if churn.len() > churn_threshold {
+                warn!("Churn larger than threshold limit!");
+                let number_to_remove = churn.len() - churn_threshold;
+
+                let elements_to_remove: Vec<&RsaIdentity> = churn.choose_multiple(
+                    &mut rand::thread_rng(),
+                    number_to_remove
+                ).collect();
+
+                for element in elements_to_remove {
+                    churn_set.remove(element);
+                }
+            }
+
+            // We remove the churned routers from the consensus.
+            unvalidated.consensus.routers
+                .retain(|r| !churn_set.contains(r.rsa_identity()));
+        }
```

Comment:
We still need this code here, otherwise, the churned relays will not be removed from the un-validated consensus.
```
unvalidated.consensus.routers.retain(|r| !churn_set.contains(r.rsa_identity()));
```

Reply:
We completely agree! :-)